src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorDocumentMappingService.cs

```diff
@@ -18,6 +18,73 @@ namespace Microsoft.AspNetCore.Razor.LanguageServer
 {
     internal class DefaultRazorDocumentMappingService : RazorDocumentMappingService
     {
+        public override bool TryMapFromProjectedDocumentEdit(RazorCodeDocument codeDocument, TextEdit edit, out TextEdit newEdit)
+        {
+            newEdit = default;
+
+            var csharpSourceText = codeDocument.GetCSharpSourceText();
+            var range = edit.Range;
+            if (!IsRangeWithinDocument(range, csharpSourceText))
+            {
+                return false;
+            }
+
+            var startIndex = range.Start.GetAbsoluteIndex(csharpSourceText);
+            var endIndex = range.End.GetAbsoluteIndex(csharpSourceText);
+            var mappedStart = TryMapFromProjectedDocumentPosition(codeDocument, startIndex, out var hostDocumentStart, out _);
+            var mappedEnd = TryMapFromProjectedDocumentPosition(codeDocument, endIndex, out var hostDocumentEnd, out _);
+
+            // Ideal case, both start and end can be mapped so just return the edit
+            if (mappedStart && mappedEnd)
+            {
+                newEdit = new TextEdit()
+                {
+                    NewText = edit.NewText,
+                    Range = new Range(hostDocumentStart, hostDocumentEnd)
+                };
+                return true;
+            }
+
+            // For the first line of a code block the C# formatter will often return an edit that starts
+            // before our mapping, but ends within. In those cases, when the edit spans multiple lines
+            // we just take the last line and try to use that.
+            //
+            // eg in the C# document you might see:
+            //
+            //      protected override void BuildRenderTree(Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder __builder)
+            //      {
+            // #nullable restore
+            // #line 1 "/path/to/Document.component"
+            //    
+            //          var x = DateTime.Now;
+            //
+            // To indent the 'var x' line the formatter will return an edit that starts the line before,
+            // with a NewText of '\n            '. The start of that edit is outside our mapping, but we
```

Comment:
>             // with a NewText of '\n            '. The start of that edit is outside our mapping, but we

Interesting, is there value in ensuring that the "first edit" effectively encapsulates only whitespace on an isolated line?